File: resources/GeerdeteBox_Teil2/Output.cpp
#include "Output.h"

// saves output in VTK format
void Output::fields(World& world, std::vector<Species>& species) {
	std::stringstream name;
	name << "results\\fields_" << std::setfill('0') << std::setw(5) << world.getTs() << ".vti";	// Windows
	//name << "results/fields_" << std::setfill('0') << std::setw(5) << world.getTs() << ".vti";// Linux (others)
	
	// open output file
	std::ofstream out(name.str());
	if (!out.is_open()) { std::cerr << "Could not open " << name.str() << std::endl; return; }

	// ImageData is a VTK format for structured Cartesian meshes
	out << "<VTKFile type=\"ImageData\">\n";
	double3 x0 = world.getX0();
	double3 dh = world.getDh();
	out << "<ImageData Origin=\"" << x0[0] << " " << x0[1] << " " << x0[2] << "\" ";
	out << "Spacing=\"" << dh[0] << " " << dh[1] << " " << dh[2] << "\" ";
	out << "WholeExtent=\"0 " << world.ni - 1 << " 0 " << world.nj - 1 << " 0 " << world.nk - 1 << "\">\n";

	// output data stored on nodes (point data)
	out << "<PointData>\n";

	// node volumes, scalar
	out << "<DataArray Name=\"NodeVol\" NumberOfComponents=\"1\" format=\"ascii\" type=\"Float64\">\n";
	out << world.node_vol;	// use the overloaded << operator
	out << "</DataArray>\n";

	// potential, scalar
	out << "<DataArray Name=\"phi\" NumberOfComponents=\"1\" format=\"ascii\" type=\"Float64\">\n";
	out << world.phi;
	out << "</DataArray>\n";

	// charge density, scalar
	out << "<DataArray Name=\"rho\" NumberOfComponents=\"1\" format=\"ascii\" type=\"Float64\">\n";
	out << world.rho;
	out << "</DataArray>\n";

	// electric field, 3 component vector
	out << "<DataArray Name=\"ef\" NumberOfComponents=\"3\" format=\"ascii\" type=\"Float64\">\n";
	out << world.ef;	// uses overloaded << from Field_ and vec3
	out << "</DataArray>\n";

	// species number densities
	for (Species& sp : species) {
		out << "<DataArray Name=\"nd." << sp.name << "\" NumberOfComponents=\"1\" format=\"ascii\" type=\"Float64\">\n";
		out << sp.den;
		out << "</DataArray>\n";
	}

	// close the tags
	out << "</PointData>\n";
	out << "</ImageData>\n";
	out << "</VTKFile>\n";
	out.close();
}

// writes information to the screen
void Output::screenOutput(World& world, std::vector<Species>& species) {
	std::cout << "ts: " << world.getTs();
	for (Species& sp : species)
		std::cout << "\t " << sp.name << ": " << sp.getNp();
	std::cout << std::endl;
}

namespace Output { std::ofstream f_diag; }	// file handle
void Output::diagOutput(World& world, std::vector<Species>& species) {
	using namespace Output;	// to get acces to f_diag
	if (!f_diag.is_open()) {			// if file not open
		f_diag.open("runtime_diags.csv");
		f_diag << "ts,time,wall_time";	// write header
		for (Species& sp : species)		// species specific header
			f_diag << ",mp_count." << sp.name << ",real_count." << sp.name << ",px." << sp.name << ",py." << sp.name << ",pz." << sp.name << ",KE." << sp.name;
		f_diag << ",PE,total_E" << std::endl;
	}

	f_diag << world.getTs() << "," << world.getTime();
	f_diag << "," << world.getWallTime();

	// write out species kinetic energy and momentum
	double tot_KE = 0;
	for (Species& sp : species) {
		double KE = sp.getKE();	// species kinetic energy
		tot_KE += KE;			// accumulate total kinetic energy
		double3 mom = sp.getMomentum();	// momentum

		f_diag << "," << sp.getNp() << "," << sp.getRealCount() << "," << mom[0] << "," << mom[1] << "," << mom[2] << "," << KE;
	}

	// write out total potential and kinetic energy
	double PE = world.getPE();
	f_diag << "," << PE << "," << (PE + tot_KE);

	f_diag << "\n";	// use \n to avoid flush to disc
	if (world.getTs() % 10 == 0) f_diag.flush();	// periodically write
}

File: testing_scripts/Sphere/output.py
import os
import numpy as np
from datetime import datetime
from world import *
class Output:

    @staticmethod
    def create_results_dir(output_dir="results"):
        try:
            os.makedirs(output_dir, exist_ok=True)
            # print(f"Directory '{output_dir}' created successfully.")
        except OSError as error:
            return 0
            # print(f"Failed to create directory '{output_dir}': {error}")

    @staticmethod
    def fields(world, species_list, filename_prefix="fields",project_dir=""):
        # The output directory is predefined as 'results'
        output_dir = os.path.join(project_dir, "results")
        
        # Ensure the output directory exists
        Output.create_results_dir(output_dir)
        
        timestamp = datetime.now().strftime("%d_%m_%Y_%H%M%S")

        # Construct the filename using the timestep
        filename = os.path.join(output_dir, f"{filename_prefix}_{timestamp}.vti")
        
        try:
            with open(filename, 'w') as out:
                # Write the VTK header for ImageData
                out.write('<VTKFile type="ImageData">\n')
                
                x0 = world.get_x0()
                dh = world.get_dh()

                # Write the ImageData attributes
                out.write(f'<ImageData Origin="{x0[0]} {x0[1]} {x0[2]}" ')
                out.write(f'Spacing="{dh[0]} {dh[1]} {dh[2]}" ')
                out.write(f'WholeExtent="0 {world.ni - 1} 0 {world.nj - 1} 0 {world.nk - 1}">\n')

                # Start PointData section
                out.write('<PointData>\n')

                # Write node_vol data
                out.write('<DataArray Name="NodeVol" NumberOfComponents="1" format="ascii" type="Float64">\n')
                out.write(Output.field_data_to_string(world.node_vol))
                out.write('</DataArray>\n')

                # Write phi data
                out.write('<DataArray Name="phi" NumberOfComponents="1" format="ascii" type="Float64">\n')
                out.write(Output.field_data_to_string(world.phi))
                out.write('</DataArray>\n')

                # Write rho data
                out.write('<DataArray Name="rho" NumberOfComponents="1" format="ascii" type="Float64">\n')
                out.write(Output.field_data_to_string(world.rho))
                out.write('</DataArray>\n')

                # Write ef data (vector field)
                out.write('<DataArray Name="ef" NumberOfComponents="3" format="ascii" type="Float64">\n')
                out.write(Output.vector_field_data_to_string(world.ef))
                out.write('</DataArray>\n')

                # Write species number densities
                for sp in species_list:
                    out.write(f'<DataArray Name="nd.{sp.name}" NumberOfComponents="1" format="ascii" type="Float64">\n')
                    out.write(Output.field_data_to_string(sp.den))
                    out.write('</DataArray>\n')

                # Close tags
                out.write('</PointData>\n')
                out.write('</ImageData>\n')
                out.write('</VTKFile>\n')

            #print(f"Data successfully written to {filename}")

        except IOError as e:
            print(f"Error writing to file {filename}: {e}")
            
    def fields_without_species(world, filename_prefix, project_dir=""):
    # The output directory is predefined as 'results'
        output_dir = os.path.join(project_dir, "results")
        
        # Ensure the output directory exists
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
    
        # Create a timestamp for the filename
        timestamp = datetime.now().strftime("%d_%m_%Y_%H%M%S")
    
        # Construct the filename using the prefix and timestamp
        filename = os.path.join(output_dir, f"{filename_prefix}_{timestamp}.vti")
    
        try:
            with open(filename, 'w') as out:
                # Write the VTK header for ImageData
                out.write('<VTKFile type="ImageData">\n')
                
                x0 = world.get_x0()
                dh = world.get_dh()
    
                # Write the ImageData attributes
                out.write(f'<ImageData Origin="{x0[0]} {x0[1]} {x0[2]}" ')
                out.write(f'Spacing="{dh[0]} {dh[1]} {dh[2]}" ')
                out.write(f'WholeExtent="0 {world.ni - 1} 0 {world.nj - 1} 0 {world.nk - 1}">\n')
    
                # Start PointData section
                out.write('<PointData>\n')
    
                # Write node_vol data
                out.write('<DataArray Name="NodeVol" NumberOfComponents="1" format="ascii" type="Float64">\n')
                out.write(Output.field_data_to_string(world.node_vol))
                out.write('</DataArray>\n')
    
                # Write phi data
                out.write('<DataArray Name="phi" NumberOfComponents="1" format="ascii" type="Float64">\n')
                out.write(Output.field_data_to_string(world.phi))
                out.write('</DataArray>\n')
    
                # Write rho data
                out.write('<DataArray Name="rho" NumberOfComponents="1" format="ascii" type="Float64">\n')
                out.write(Output.field_data_to_string(world.rho))
                out.write('</DataArray>\n')
    
                # Write ef data (vector field)
                out.write('<DataArray Name="ef" NumberOfComponents="3" format="ascii" type="Float64">\n')
                out.write(Output.vector_field_data_to_string(world.ef))
                out.write('</DataArray>\n')
    
                # Close tags
                out.write('</PointData>\n')
                out.write('</ImageData>\n')
                out.write('</VTKFile>\n')
    
            print(f"Data successfully written to {filename}")
    
        except Exception as e:
            print(f"Could not write data to {filename}: {e}")
    
        
    @staticmethod
    def screen_output(world, species_list):
        output = f"ts: {world.get_ts()}"
        for sp in species_list:
            real_count = int(round(sp.get_real_count()))  # Ensure it's an integer
            output += f"  {sp.name}: {real_count}"
        print(output)

    @staticmethod
    def diag_output(world, species_list, project_dir=""):
        # print(f"Iteration: {world.get_ts()}")
        diag_file = os.path.join(project_dir, "runtime_diags.csv")
        
        # Check if the file exists to write the header
        write_header = not os.path.exists(diag_file)
        
        with open(diag_file, 'a') as f_diag:
            if write_header:
                # Write header
                header = "ts,time,wall_time"
                for sp in species_list:
                    header += f",mp_count.{sp.name},real_count.{sp.name},px.{sp.name},py.{sp.name},pz.{sp.name},KE.{sp.name}"
                header += ",PE,total_E\n"
                f_diag.write(header)
            
            # Write data
            line = f"{world.get_ts()},{world.get_time()},{world.get_wall_time()}"
            
            total_KE = 0
            for sp in species_list:
                KE = sp.get_ke()
                total_KE += KE
                mom = sp.get_momentum()
                line += f",{sp.get_np()},{sp.get_real_count()},{mom[0]},{mom[1]},{mom[2]},{KE}"
            
            PE = world.get_pe()
            total_E = PE + total_KE
            line += f",{PE},{total_E}\n"
            f_diag.write(line)
        
        # # Print the energy values to the console
        # print(f"Total KE: {total_KE}")
        # print(f"PE: {PE}")
        # print(f"Total Energy: {total_E}")

       
    # @staticmethod
    # def diag_output(world, species_list, project_dir=""):
    #     diag_file = os.path.join(project_dir, "runtime_diags.csv")
        
    #     # Check if the file exists to write the header
    #     write_header = not os.path.exists(diag_file)
        
    #     with open(diag_file, 'a') as f_diag:
    #         if write_header:
    #             # Write header
    #             header = "ts,time,wall_time"
    #             for sp in species_list:
    #                 header += f",mp_count.{sp.name},real_count.{sp.name},px.{sp.name},py.{sp.name},pz.{sp.name},KE.{sp.name}"
    #             header += ",PE,total_E\n"
    #             f_diag.write(header)
            
    #         # Write data
    #         line = f"{world.get_ts()},{world.get_time()},{world.get_wall_time()}"
            
    #         total_KE = 0
    #         for sp in species_list:
    #             KE = sp.get_ke()
    #             total_KE += KE
    #             mom = sp.get_momentum()
    #             line += f",{sp.get_np()},{sp.get_real_count()},{mom[0]},{mom[1]},{mom[2]},{KE}"
            
    #         PE = world.get_pe()
    #         line += f",{PE},{PE + total_KE}\n"
    #         f_diag.write(line)

    @staticmethod
    def field_data_to_string(field):
        """Converts scalar field data to a string in VTK format."""
        data_str = ""
        for k in range(field.nk):
            for j in range(field.nj):
                for i in range(field.ni):
                    data_str += f"{field(i, j, k)} "
        return data_str + "\n"

    @staticmethod
    def vector_field_data_to_string(field):
        """Converts vector field data to a string in VTK format."""
        data_str = ""
        for k in range(field.nk):
            for j in range(field.nj):
                for i in range(field.ni):
                    data_str += f"{field(i, j, k, 0)} {field(i, j, k, 1)} {field(i, j, k, 2)} "
        return data_str + "\n"
